### ISSUE TITLE
Support for multiple quadratures in a kernel

### ISSUE THREAD

**Issue by thomasmelvin** [stfc/PSyclone#195]
Some of the kernels in LFRic [#1449](https://code.metoffice.gov.uk/trac/lfric/ticket/1449) use multiple (2) quadratures. This is so that they can perform both 3d quadrature in the volume and also 2d (or 1d) quadrature on faces (or edges).

The change suggested in LFRic [#1449](https://code.metoffice.gov.uk/trac/lfric/ticket/1449) is to add a second `gh_shape` to give:

```fortran
  integer :: gh_shape  = GH_QUADRATURE_XYoZ
  integer :: gh_shape2 = GH_QUADRATURE_face
```

See for example `poly1d_flux_coeffs_kernel_mod.F90`

**Comment by arporter** [comment]
How about:

    integer(2) :: gh_shape = (/GH_QUADRATURE_XYoZ, GH_QUADRATURE_face/)

**Comment by thomasmelvin** [comment]
That looks good to me

**Comment by TeranIvy** [comment]
Does this mean that a kernel that requires support for just one quadrature type would still have `gh_shape` defined as an integer array of size 1? If so, we need to adjust the LFRic kernels once this change becomes a part of a new PSyclone release.

**Comment by arporter** [comment]
No, I was thinking that if there's only one shape then gh_shape can be a scalar. Might be better illustrated as:

    integer :: gh_shape(2) = (/.../)

and

    integer :: gh_shape = GH_QUADRATURE_XYoZ

(I think that's valid Fortran?)

**Comment by arporter** [comment]
I'm going to add support for this as part of #193 but I have a question for @thomasmelvin or @TeranIvy - would it ever make sense for a kernel to request an evaluator *and* quadrature, e.g.:

    gh_shape(2) = (\ gh_quadrature_face, gh_evaluator \)

or should we forbid such things?

**Comment by thomasmelvin** [comment]
> I'm going to add support for this as part of #193 but I have a question for @thomasmelvin or @TeranIvy - would it ever make sense for a kernel to request an evaluator _and_ quadrature, e.g.:
> 
> ```
> gh_shape(2) = (\ gh_quadrature_face, gh_evaluator \)
> ```
> 
> or should we forbid such things?

Yes, makes perfect sense to me

**Comment by arporter** [comment]
> 
> 
> > I'm going to add support for this as part of #193 but I have a question for @thomasmelvin or @TeranIvy - would it ever make sense for a kernel to request an evaluator _and_ quadrature, e.g.:
> > ```
> > gh_shape(2) = (\ gh_quadrature_face, gh_evaluator \)
> > ```
> > 
> > 
> > or should we forbid such things?
> 
> Yes, makes perfect sense to me

As in, it makes sense to support it?

**Comment by thomasmelvin** [comment]
> > > I'm going to add support for this as part of #193 but I have a question for @thomasmelvin or @TeranIvy - would it ever make sense for a kernel to request an evaluator _and_ quadrature, e.g.:
> > > ```
> > > gh_shape(2) = (\ gh_quadrature_face, gh_evaluator \)
> > > ```
> > > 
> > > 
> > > or should we forbid such things?
> > 
> > 
> > Yes, makes perfect sense to me
> 
> As in, it makes sense to support it?

Sorry, I should have been more explicit,

Yes, I think we should support it.  I can easily imagine a scenario where we want to locally compute a field (i.e. sample on nodal points using an evaluator) and then integrate this field over the cell (using a quadrature).

To be complete I think we should support:
`gh_shape(2) = (\ gh_quadrature_1, gh_quadrature_2 \)`
`gh_shape(2) = (\ gh_quadrature, gh_evaluator \)`
`gh_shape(2) = (\ gh_evaluator_1, gh_evaluator_2 \)`

**Comment by arporter** [comment]
Thanks @thomasmelvin. Just to be clear, the only way of specifying that an evaluator is required is currently to have a gh_shape of `gh_evaluator`.  Therefore, I don't know how we would support your third example? Are we missing a concept in the meta-data?

**Comment by thomasmelvin** [comment]
I hadn't considered this. I guess the evaluator is always assumed to be evaluated at the nodal points of the first field? 

I guess we would need to extend the metadata? something like
```fortran 
   type, public, extends(kernel_type) :: some_kernel_type
    private
    type(arg_type) :: meta_args(2) = (/            &
        arg_type(GH_FIELD,   GH_INC,  W1),      &
        arg_type(GH_FIELD,   GH_READ, W2)     &
        /)
    type(func_type) :: meta_funcs(2) = (/ &
        func_type(W1, GH_BASIS),             &
        func_type(W2, GH_DIFF_BASIS)     &
        /)
    integer :: iterates_over = CELLS
    integer :: gh_shape(2) = (/ GH_EVALUATOR_W1, GH_EVALUATOR_W2 /)
  contains
    procedure, nopass :: some_code
  end type
```

But if this is not straightforward to do then I don't suggest it is implemented unless we have a strong need for it

**Comment by arporter** [comment]
I've just read our documentation and realise that in fact we do already support multiple evaluators! You simply specify a list of target function spaces in `gh_evaluator_targets`. So, I think we're OK...

**Comment by TeranIvy** [comment]
Closed via #701
